pandas/_libs/tslibs/parsing.pyx

```diff
@@ -44,13 +50,73 @@ class DateParseError(ValueError):
 
 _DEFAULT_DATETIME = datetime(1, 1, 1).replace(hour=0, minute=0,
                                               second=0, microsecond=0)
+_DEFAULT_TZINFO = _DEFAULT_DATETIME.tzinfo
 
 cdef:
     object _TIMEPAT = re.compile(r'^([01]?[0-9]|2[0-3]):([0-5][0-9])')
 
     set _not_datelike_strings = {'a', 'A', 'm', 'M', 'p', 'P', 't', 'T'}
 
 # ----------------------------------------------------------------------
+DEF delimiters = b' /-\\'
+DEF MAX_DAYS_IN_MONTH = 31
+DEF MAX_MONTH = 12
+
+cdef bint _is_not_delimiter(const char ch):
+    return strchr(delimiters, ch) == NULL
+
+cdef inline int _parse_2digit(const char* s):
+    cdef int result = 0
+    result += getdigit_ascii(s[0], -10) * 10
+    result += getdigit_ascii(s[1], -100) * 1
+    return result
+
+cdef inline int _parse_4digit(const char* s):
+    cdef int result = 0
+    result += getdigit_ascii(s[0], -10) * 1000
+    result += getdigit_ascii(s[1], -100) * 100
+    result += getdigit_ascii(s[2], -1000) * 10
+    result += getdigit_ascii(s[3], -10000) * 1
+    return result
+
+cdef inline object parse_delimited_date(object date_string, bint dayfirst,
+                                        object tzinfo):
+    cdef:
+        const char* buf
+        Py_ssize_t length
+        int day = 1, month = 1, year
+
+    assert isinstance(date_string, (str, unicode))
+    buf = get_c_string_buf_and_size(date_string, &length)
+    if length == 10:
+        if _is_not_delimiter(buf[2]) or _is_not_delimiter(buf[5]):
+            return None, None
+        month = _parse_2digit(buf)
+        day = _parse_2digit(buf + 3)
+        year = _parse_4digit(buf + 6)
+        reso = 'day'
+    elif length == 7:
+        if _is_not_delimiter(buf[2]):
+            return None, None
+        month = _parse_2digit(buf)
+        year = _parse_4digit(buf + 3)
+        reso = 'month'
+    else:
+        return None, None
+
+    if month < 0 or day < 0 or year < 0:
+        # some part is not an integer, so it's not a mm/dd/yyyy date
+        return None, None
+
+    if 1 <= month <= MAX_DAYS_IN_MONTH and 1 <= day <= MAX_DAYS_IN_MONTH \
+            and (month <= MAX_MONTH or day <= MAX_MONTH):
+        if month > MAX_MONTH or (day < MAX_MONTH and dayfirst):
+            day, month = month, day
+        return PyDateTimeAPI.DateTime_FromDateAndTime(
```

Comment:
a fairly simple optimization you could push throughout _libs would be to use the PyDatetimeAPI constructors consistently; we currently just call `datetime(...)` everywhere, which im pretty sure uses the python constructor (though it seems like cython could make that optimization on its own...)

Reply:
`datetime` call uses `datetime_new()` Python function. The performance issue with it is it expects it parameters as a Python `tuple` of Python `int`-s, which takes a lot of time to convert Cython `int`-s to it (and then `datetime` would unpack it back to C `int`-s). All this back-and-forth usually costs more than actually constructing the datetime object.

We can make a separate PR that replaces `datetime()` with this call, yes :)

Reply:
yes pls do this separately, again use the idioms that are already here.

Reply:
Okay, rough measurements show that `datetime()` is about 6-8% slower than C api calls:

date format|C api|`datetime()`
-|-|-
mY|46.6 ms|50.9 ms
mdY|18.4 ms|19.5 ms

We can definitely replace `datetime()` with C api everywhere in a separate PR.

Reply:
I have re-measured with Cython inlined function `datetime_new` and I see no different with C api calls, so I will change our PR to use that function instead.